affinity/cpp-23/dXXXXr0.md

```diff
@@ -0,0 +1,295 @@
+# DXXXXr0: Policy property for describing adjacency
+
+**Date: 2020-04-06**
+
+**Audience: SG1, SG14**
+
+**Authors: Gordon Brown, Ruyman Reyes, Michael Wong, H. Carter Edwards, Thomas
+Rodgers, Mark Hoemmen, Tom Scogland**
+
+**Emails: gordon@codeplay.com, ruyman@codeplay.com, michael@codeplay.com,
+hedwards@nvidia.com, rodgert@twrodgers.com, mhoemme@sandia.gov,
+tscogland@llnl.gov**
+
+**Reply to: gordon@codeplay.com**
+
+# Acknowledgements
+
+This paper is the result of discussions from many contributors within SG1, SG14
+and the heterogeneous C\+\+ group.
+
+# Preface
+
+After a lengthy discussion with SG1 during the Prague 2019 ISO C++ meeting it
+was decided to take a different approach to P1436r3 [[1]][P1436r3]. This paper
+is a continuation of prior work in P1436r3 [[2]][p0443r12] focusing specifically
+on the the `adjacency` properties, previously named `bulk_execution_affinity`.
+
+# Motivation
+
+*Affinity* refers to the "closeness" in terms of memory access performance,
+between running code, the hardware execution resource on which the code runs,
+and the data that the code accesses.  A hardware execution resource has "more
+affinity" to a part of memory or to some data, if it has lower latency and/or
+higher bandwidth when accessing that memory / those data.
+
+On almost all computer architectures, the cost of accessing different data may
+differ. Most computers have caches that are associated with specific processing
+units. If the operating system moves a thread or process from one processing
+unit to another, the thread or process will no longer have data in its new cache
+that it had in its old cache. This may make the next access to those data
+slower. Many computers also have a Non-Uniform Memory Architecture (NUMA),
+which means that even though all processing units see a single memory in terms
+of programming model, different processing units may still have more affinity to
+some parts of memory than others. NUMA exists because it is difficult to scale
+non-NUMA memory systems to the performance needed by today's highly parallel
+computers and applications.
+
+One strategy to improve applications' performance, given the importance of
+affinity, is processor and memory *binding*. Keeping a process bound to a
+specific thread and local memory region optimizes cache affinity. It also
+reduces context switching and unnecessary scheduler activity. Since memory
+accesses to remote locations incur higher latency and/or lower bandwidth,
+control of thread placement to enforce affinity within parallel applications is
```

Comment:
"Thread placement" first shows up italicized in the next paragraph, but we use it here.  Do we want to say that placement is the same as binding?

Reply:
What I'm getting at, is that threads might get an initial placement, but may or may not be allowed to migrate around from there.  For instance, it could make sense to assign threads initially to certain cores, but permit them to migrate for better load balancing or to avoid overheating.  I see binding as a special case of placement.

Reply:
That's a good point, I've reworked the motivation slightly in my most recent change, to better reflect the new approach. I've tried to focus on the term "thread placement" as this seems like the more general term, while "binding threads" is, as you said, a special case of this.